### PR TITLE
Implement QK attention head chunking in MLA to reduce HBM footprint

### PR DESCRIPTION
**Pull request by zcjhao**
# Description

Currently, the QK dot product natively in MLA materializes the full `[batch, count_of_heads, q_len, kv_len]` attention scores tensor. This causes memory constraints for large context lengths. 
  
  This change introduces the Config variable `qk_head_chunk_size` alongside a `jax.lax.scan` algorithm. Since the `heads` dimension is locally unsharded (unlike `q_len` which is Context Parallelized), we scan over groups of heads iteratively over the QK dimension to compute query-key projections and their softmax partials.

  If `qk_head_chunk_size` is left at the default `0`, the attention layer automatically falls back to the native, un-
chunked tensor multiplication.

# Tests

  - Updated `tests/unit/deepseek32_vs_reference_test.py::DeepseekV32IndexerTest::test_indexer_match` to automatically sweep multiple chunk sizes (`qk_head_chunk_size` = `0`,  `2`, and `4`).
  - Added a chunked test case (`qk_head_chunk_size=4`) to `DeepseekV32MLATest::test_mla_parity` to verify mathematical equivalence with the PyTorch reference.
  
  Test output: 
  
  ```
      ============================= test session starts ==============================
    platform linux -- Python 3.13.14, pytest-9.0.3
    rootdir: /usr/local/google/home/zjiahao/maxtext-qk-chunk
    configfile: pytest.ini
    collected 10 items
    
    tests/unit/deepseek32_vs_reference_test.py::DeepseekV32IndexerTest::test_indexer_match0 PASSED [ 10%]
    tests/unit/deepseek32_vs_reference_test.py::DeepseekV32IndexerTest::test_indexer_match1 PASSED [ 20%]
    tests/unit/deepseek32_vs_reference_test.py::DeepseekV32IndexerTest::test_indexer_match2 PASSED [ 30%]
    tests/unit/deepseek32_vs_reference_test.py::DeepseekV32MLATest::test_mla_parity_dot_product_s128_k128 PASSED [ 40%]
    tests/unit/deepseek32_vs_reference_test.py::DeepseekV32MLATest::test_mla_parity_dot_product_s128_k128_c4 PASSED [
  50%]
    tests/unit/deepseek32_vs_reference_test.py::DeepseekV32MLATest::test_mla_parity_dot_product_s128_k4 PASSED [ 60%]
    tests/unit/deepseek32_vs_reference_test.py::DeepseekV32MLATest::test_mla_parity_dot_product_s2_k4 PASSED [ 70%]
    tests/unit/deepseek32_vs_reference_test.py::DeepseekV32MLATest::test_mla_parity_dot_product_s8_k4 PASSED [ 80%]
    tests/unit/deepseek32_vs_reference_test.py::DeepseekV32MLATest::test_mla_parity_flash_s128_k128 FAILED [ 90%]
    tests/unit/deepseek32_vs_reference_test.py::DeepseekV32MLATest::test_mla_parity_flash_s128_k4 FAILED [ 100%]
    
    =========================== short test summary info ============================
    FAILED tests/unit/deepseek32_vs_reference_test.py::DeepseekV32MLATest::test_mla_parity_flash_s128_k128
    FAILED tests/unit/deepseek32_vs_reference_test.py::DeepseekV32MLATest::test_mla_parity_flash_s128_k4
    ============= 2 failed, 8 passed, 65 warnings in 65.37s (0:01:05) ==============
  ```
  
  The two flash failures are unrelated to this PR, we do not use flash attention for QK in MLA currently. Running `python -m pytest tests/unit/deepseek32_vs_reference_test.py -k "test_mla_parity_flash" -v` would also fail in a clean workspace synced with the HEAD.
  
  -  Verified reduced memory footprints (+3s step time when using `qk_head_chunk_size=16`):
[After](https://screenshot.googleplex.com/6NLyAHd9tfU9JyN)
[Before](https://screenshot.googleplex.com/9b3UDdCH62KH27b)

# Checklist 

Before submitting this PR, please make sure (put X in square brackets):

[✓] I have performed a self-review of my code. For an optional AI review, add the gemini-review label.
[✓] I have necessary comments in my code, particularly in hard-to-understand areas.
[✓] I have run end-to-end tests tests and provided workload links above if applicable.
[✓ ] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in our documentation https://maxtext.readthedocs.
io/en/latest/development.html#adding-new-documentation-files.

### CODE REVIEW

File: src/maxtext/layers/attention_mla.py [hunk 1]
```diff
@@ -1086,25 +1124,69 @@ def calculate_indexer_loss(
...
+          key,
+          precision=self.config.matmul_precision,
+      )
+      if sparse_loss:
+        attention_scores = attention_scores + indexer_mask[:, None, :, :]
+      elif attention_mask is not None:
+        attention_scores = attention_scores + attention_mask[:, None, :, :]
+      attention_probs = jnp.sum(jax.nn.softmax(attention_scores.astype(jnp.float32), axis=-1), axis=1)
+      attention_probs = jax.lax.optimization_barrier(attention_probs)
     # L1 normalize aggregated target distribution
```
[the remark this answers is not in the document]
Comment (gobbleturk) [reply src/maxtext/layers/attention_mla.py:1187]: this looks like a reasonable recommendation - do we need the optimization barrier in the chunked case as well? Maybe XLA is doing well without it?

File: src/maxtext/layers/attention_mla.py [hunk 2]
```diff
@@ -361,18 +361,56 @@ def __call__(
...
+    # Compute Index Scores
+    # When qk_head_chunk_size > 0, compute Index Scores by chunking the 'heads' dimension to reduce memory
+    # The naive evaluation materializes [b, t, s, h].
+    # We use jax.lax.scan to compute the score iteratively over head chunks.
+    b, t, h, d = q.shape
+    # Control the HBM footprint of QK tensor: [batch, q_len, s_len, heads]
+    # If set to 0 (defaults), it falls back to native materialization.
+    head_chunk_size = getattr(self.config, "qk_head_chunk_size", 0)
+    if head_chunk_size > 0:
+      if head_chunk_size > h or h % head_chunk_size != 0:
```
Comment (gobbleturk) [inline src/maxtext/layers/attention_mla.py]: Consider wrapping lines 379 to 407 in a function chunked_head_loop (identical code but wrapped in a function), following code style go/small-functions

**Review by gobbleturk — APPROVED** [review]

File: src/maxtext/layers/attention_mla.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zcjhao) [reply src/maxtext/layers/attention_mla.py:1187]: Mmmm. Let me check

[the remark this answers is not in the document]
Comment (zcjhao) [reply src/maxtext/layers/attention_mla.py:1187]: no barrier needed for chunked case! :)

File: src/maxtext/layers/attention_mla.py (hunk 2 above)
Comment (zcjhao) [reply src/maxtext/layers/attention_mla.py]: Done!

File: src/maxtext/configs/base.yml [hunk 3]
```diff
@@ -1339,3 +1336,6 @@ elastic_enabled: false
 elastic_timeout_seconds: 300
 elastic_max_retries: 10
 elastic_min_slice_count: -1
+
+# Limits HBM footprint by sequentially evaluating the QK matrix across the unsharded local heads dimension natively. Size must evenly divide the number of attention heads.
+qk_head_chunk_size: 0
```
Comment (RissyRan) [inline src/maxtext/configs/base.yml]: Shall we move this config near attn? For instance, like here: https://github.com/AI-Hypercomputer/maxtext/blob/7b0e7980ed5d601848a70a315f82f90fedf39acf/src/maxtext/configs/base.yml#L194

File: src/maxtext/configs/types.py [hunk 4]
```diff
@@ -637,6 +637,16 @@ class Attention(BaseModel):
   force_q_layout: bool = Field(False, description="Force the Q layout")
   use_qk_clip: bool = Field(False, description="Whether to use QK-Clip (MuonClip) for training stability.")
   qk_clip_threshold: float = Field(100.0, description="Threshold for QK-Clip (tau).")
+  qk_head_chunk_size: int = Field(
```
Comment (RissyRan) [inline src/maxtext/configs/types.py]: Let's add an assertion for this `https://github.com/AI-Hypercomputer/maxtext/blob/7b0e7980ed5d601848a70a315f82f90fedf39acf/src/maxtext/configs/base.yml#L194` and remove context here. 

One example in this types.py:

https://github.com/AI-Hypercomputer/maxtext/blob/7b0e7980ed5d601848a70a315f82f90fedf39acf/src/maxtext/configs/types.py#L3264

File: src/maxtext/layers/attention_mla.py [hunk 5]
```diff
@@ -361,18 +361,60 @@ def __call__(
...
+    # When qk_head_chunk_size > 0, compute Index Scores by chunking the 'heads' dimension to reduce memory
+    # The naive evaluation materializes [b, t, s, h].
+    # We use jax.lax.scan to compute the score iteratively over head chunks.
+    b, t, h, d = q.shape
+    # Control the HBM footprint of QK tensor: [batch, q_len, s_len, heads]
+    # If set to 0 (defaults), it falls back to native materialization.
+    head_chunk_size = getattr(self.config, "qk_head_chunk_size", 0)
+
+    def chunked_head_loop(chunk_size):
+      if chunk_size > h or h % chunk_size != 0:
```
Comment (RissyRan) [inline src/maxtext/layers/attention_mla.py]: Let's move the assertion for early check in types.py?

File: src/maxtext/layers/attention_mla.py [hunk 6]
```diff
@@ -1086,25 +1128,69 @@ def calculate_indexer_loss(
...
-    # Force materialization and prevent fusion across this point to reuse the intermediate tensor
-    attention_probs = jax.lax.optimization_barrier(attention_probs)
+    batch, q_len, heads, dim = query.shape
+
+    # Chunk across the 'heads' dimension manually using jax.lax.scan
+    # Control the HBM footprint of QK tensor: [batch, q_len, s_len, heads]
+    # If set to 0, it falls back to native implementation.
+    head_chunk_size = getattr(self.config, "qk_head_chunk_size", 0)
+    if head_chunk_size > 0:
+      if head_chunk_size > heads or heads % head_chunk_size != 0:
```
Comment (RissyRan) [inline src/maxtext/layers/attention_mla.py]: Shall we have a helper function and re-use?

**Review by RissyRan — COMMENTED** [review]
Thanks for the change! Minor comments. 

Also, I don't recall if we have auto test for deepseek32_vs_reference_tests. Could you attach your local test result in PR description? Thanks!

File: src/maxtext/configs/base.yml (hunk 3 above)
Comment (zcjhao) [reply src/maxtext/configs/base.yml]: You are right, I see in the [GIthub workflow file](https://github.com/AI-Hypercomputer/maxtext/blob/7b0e7980ed5d601848a70a315f82f90fedf39acf/.github/workflows/run_tests_coordinator.yml#L212) that the deepseeeks tests are not included. Let me rerun and add the outputs in the PR.

Comment (zcjhao) [reply src/maxtext/configs/base.yml]: Moved and renamed `qk_head_chunk_size` to `mla_qk_head_chunk_size` !

File: src/maxtext/configs/types.py (hunk 4 above)
Comment (zcjhao) [reply src/maxtext/configs/types.py]: added and shortened description

File: src/maxtext/layers/attention_mla.py (hunk 6 above)
Comment (zcjhao) [reply src/maxtext/layers/attention_mla.py]: since we moved the assertion I think we do not need anymore?

File: src/maxtext/layers/attention_mla.py (hunk 5 above)
Comment (zcjhao) [reply src/maxtext/layers/attention_mla.py]: done!

File: src/maxtext/configs/base.yml [hunk 7]
```diff
@@ -1338,3 +1337,5 @@ elastic_enabled: false
 elastic_timeout_seconds: 300
 elastic_max_retries: 10
 elastic_min_slice_count: -1
+
```
Comment (RissyRan) [inline src/maxtext/configs/base.yml:1342]: Nit: let's remove.

**Review by RissyRan — APPROVED** [review]
Thanks for the change! Also, let's create a bug to track the failing the test, feel free to assign to me first.
